# First Data Set: chipotle.tsv

In [1]:
import pandas as pd

In [93]:
chipotle_data = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv", sep = "\t")

In [4]:
#head command to see first 5 records
chipotle_data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [5]:
#tail command to see first 5 records from bottom
chipotle_data.tail()

order_id  quantity           item_name  \
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   
4620      1834         1  Chicken Salad Bowl   
4621      1834         1  Chicken Salad Bowl   

                                     choice_description item_price  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75

In [13]:
#dimenion of chipotle data set
chipotle_data.shape

(4622, 5)

##### 4622 rows with 5 columns

In [17]:
#name of columns present
chipotle_data.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

In [7]:
#describe command to see statistical summary
chipotle_data.describe()  #by default it only give numeric data statistical summary

order_id     quantity
count  4622.000000  4622.000000
mean    927.254868     1.075725
std     528.890796     0.410186
min       1.000000     1.000000
25%     477.250000     1.000000
50%     926.000000     1.000000
75%    1393.000000     1.000000
max    1834.000000    15.000000

In [8]:
#for categorical data statistical summary
chipotle_data.describe(include = "object")

item_name choice_description item_price
count           4622               3376       4622
unique            50               1043         78
top     Chicken Bowl        [Diet Coke]     $8.75 
freq             726                134        730

In [12]:
#to check data types of features
chipotle_data.dtypes

order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

In [15]:
#check null values present in data set
chipotle_data.isnull().sum()

order_id                 0
quantity                 0
item_name                0
choice_description    1246
item_price               0
dtype: int64

##### from above we can see that only featute *choice_description* has 1246 null values present

In [16]:
#check percentage of null values present in every column
chipotle_data.isnull().mean()

order_id              0.00000
quantity              0.00000
item_name             0.00000
choice_description    0.26958
item_price            0.00000
dtype: float64

##### 26.958% of choice_description column has null value

### Printing those rows which has null value present in choice_description column

In [105]:
choice_null = chipotle_data[chipotle_data["choice_description"].isnull() == True]

In [106]:
choice_null.index

Int64Index([   0,    3,    6,   10,   14,   15,   20,   25,   30,   32,
            ...
            4582, 4584, 4588, 4596, 4598, 4600, 4605, 4613, 4614, 4616],
           dtype='int64', length=1246)

In [107]:
chipotle_data.iloc[choice_null.index]

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
6            3         1                          Side of Chips   
10           5         1                    Chips and Guacamole   
14           7         1                    Chips and Guacamole   
...        ...       ...                                    ...   
4600      1827         1                    Chips and Guacamole   
4605      1828         1                    Chips and Guacamole   
4613      1831         1                                  Chips   
4614      1831         1                          Bottled Water   
4616      1832         1                    Chips and Guacamole   

     choice_description item_price total_ordered_value  
0                   NaN       2.39                2.39  
3                   NaN       2.39                2.39  
6                   NaN       1.69                1.69  
10                  NaN       4.45                4.45  
14                  NaN       4.45                4.45  
...                 ...        ...                 ...  
4600                NaN       4.45                4.45  
4605                NaN       4.45                4.45  
4613                NaN       2.15                2.15  
4614                NaN       1.50                1.50  
4616                NaN       4.45                4.45  

[1246 rows x 6 columns]

In [18]:
#info of data
chipotle_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


In [33]:
#column name of feature which are categorical

col = chipotle_data.dtypes[chipotle_data.dtypes == "object"].index

col

Index(['item_name', 'choice_description', 'item_price'], dtype='object')

In [34]:
#Different categories present in item_name
chipotle_data["item_name"].value_counts()

Chicken Bowl                             726
Chicken Burrito                          553
Chips and Guacamole                      479
Steak Burrito                            368
Canned Soft Drink                        301
Steak Bowl                               211
Chips                                    211
Bottled Water                            162
Chicken Soft Tacos                       115
Chips and Fresh Tomato Salsa             110
Chicken Salad Bowl                       110
Canned Soda                              104
Side of Chips                            101
Veggie Burrito                            95
Barbacoa Burrito                          91
Veggie Bowl                               85
Carnitas Bowl                             68
Barbacoa Bowl                             66
Carnitas Burrito                          59
Steak Soft Tacos                          55
6 Pack Soft Drink                         54
Chips and Tomatillo Red Chili Salsa       48
Chicken Cr

##### item_name: Chicken Bowl is the most ordered one

In [35]:
#Different categories present in choice_description
chipotle_data["choice_description"].value_counts()

[Diet Coke]                                                                                                134
[Coke]                                                                                                     123
[Sprite]                                                                                                    77
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Lettuce]]                                      42
[Fresh Tomato Salsa, [Rice, Black Beans, Cheese, Sour Cream, Guacamole, Lettuce]]                           40
                                                                                                          ... 
[Tomatillo Green Chili Salsa, [Fajita Vegetables, Lettuce]]                                                  1
[Tomatillo Red Chili Salsa, [Cheese, Rice, Black Beans]]                                                     1
[Fresh Tomato Salsa, [Pinto Beans, Lettuce, Cheese, Rice]]                                                   1
[

In [36]:
#Different categories present in item_price
chipotle_data["item_price"].value_counts()

$8.75      730
$11.25     521
$9.25      398
$4.45      349
$8.49      311
          ... 
$7.50        1
$22.96       1
$22.20       1
$16.38       1
$33.75       1
Name: item_price, Length: 78, dtype: int64

In [39]:
chipotle_data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

### Remove  dollar sign from item_price and make it numeric to calculate:
#### total_ordered_value = item_price * quantity

In [64]:
len_item_price = len(chipotle_data["item_price"])

In [72]:
len_item_price

4622

In [94]:
for i in range(0,len_item_price):
    chipotle_data["item_price"][i] = float(chipotle_data["item_price"][i].split("$")[1])

C:\Users\USER\Anaconda3\envs\DeepLearning\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [95]:
chipotle_data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN       2.39  
1                                       [Clementine]       3.39  
2                                            [Apple]       3.39  
3                                                NaN       2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...      16.98

In [96]:
#Adding new column with name total_ordered_value 

chipotle_data["total_ordered_value"] = chipotle_data["quantity"]*chipotle_data["item_price"]

In [97]:
chipotle_data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  \
0                                                NaN       2.39   
1                                       [Clementine]       3.39   
2                                            [Apple]       3.39   
3                                                NaN       2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...      16.98   

  total_ordered_value  
0                2.39  
1                3.39  
2                3.39  
3                2.39  
4               33.96

In [110]:
#maximum ordered value
max(chipotle_data["total_ordered_value"])

663.75

### Details of maximum ordered value

In [117]:
chipotle_data["total_ordered_value"].sort_values(ascending = False)

3598    663.75
4152    150.00
1254    140.00
3602    140.00
3887    108.16
         ...  
107       1.09
195       1.09
434       1.09
2814      1.09
4069      1.09
Name: total_ordered_value, Length: 4622, dtype: object

In [125]:
#row number for maximum ordered value = 3598

chipotle_data.iloc[[3598]]

order_id  quantity                     item_name choice_description  \
3598      1443        15  Chips and Fresh Tomato Salsa                NaN   

     item_price total_ordered_value  
3598      44.25              663.75

## Filling null values with missing_record

In [127]:
chipotle_data.isnull().sum()

order_id                  0
quantity                  0
item_name                 0
choice_description     1246
item_price                0
total_ordered_value       0
dtype: int64

In [129]:
chipotle_data.fillna("missing_record", axis = 1,  inplace = True)

In [130]:
chipotle_data.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description  item_price  \
0                                     missing_record        2.39   
1                                       [Clementine]        3.39   
2                                            [Apple]        3.39   
3                                     missing_record        2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98   

   total_ordered_value  
0                 2.39  
1                 3.39  
2                 3.39  
3                 2.39  
4                33.96

# Second Data Set: us_crime_rate

#### https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/US_Crime_Rates/US_Crime_Rates_1960_2014.csv

In [143]:
us_data = pd.read_csv("https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/04_Apply/US_Crime_Rates/US_Crime_Rates_1960_2014.csv")

In [144]:
us_data.head()

Year  Population    Total  Violent  Property  Murder  Forcible_Rape  \
0  1960   179323175  3384200   288460   3095700    9110          17190   
1  1961   182992000  3488000   289390   3198600    8740          17220   
2  1962   185771000  3752200   301510   3450700    8530          17550   
3  1963   188483000  4109500   316970   3792500    8640          17650   
4  1964   191141000  4564600   364220   4200400    9360          21420   

   Robbery  Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
0   107840              154320    912100        1855400         328200  
1   106670              156760    949600        1913000         336000  
2   110860              164570    994300        2089600         366800  
3   116470              174210   1086400        2297800         408300  
4   130390              203050   1213200        2514400         472800

In [145]:
#name of columns
us_data.columns

Index(['Year', 'Population', 'Total', 'Violent', 'Property', 'Murder',
       'Forcible_Rape', 'Robbery', 'Aggravated_assault', 'Burglary',
       'Larceny_Theft', 'Vehicle_Theft'],
      dtype='object')

In [147]:
us_data.head()

Year  Population    Total  Violent  Property  Murder  Forcible_Rape  \
0  1960   179323175  3384200   288460   3095700    9110          17190   
1  1961   182992000  3488000   289390   3198600    8740          17220   
2  1962   185771000  3752200   301510   3450700    8530          17550   
3  1963   188483000  4109500   316970   3792500    8640          17650   
4  1964   191141000  4564600   364220   4200400    9360          21420   

   Robbery  Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
0   107840              154320    912100        1855400         328200  
1   106670              156760    949600        1913000         336000  
2   110860              164570    994300        2089600         366800  
3   116470              174210   1086400        2297800         408300  
4   130390              203050   1213200        2514400         472800

In [148]:
us_data.tail()

Year  Population     Total  Violent  Property  Murder  Forcible_Rape  \
50  2010   309330219  10363873  1251248   9112625   14772          85593   
51  2011   311587816  10258774  1206031   9052743   14661          84175   
52  2012   313873685  10219059  1217067   9001992   14866          85141   
53  2013   316497531   9850445  1199684   8650761   14319          82109   
54  2014   318857056   9475816  1197987   8277829   14249          84041   

    Robbery  Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
50   369089              781844   2168457        6204601         739565  
51   354772              752423   2185140        6151095         716508  
52   355051              762009   2109932        6168874         723186  
53   345095              726575   1931835        6018632         700294  
54   325802              741291   1729806        5858496         689527

In [151]:
us_data.shape

(55, 12)

In [152]:
us_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55 entries, 0 to 54
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   Year                55 non-null     int64
 1   Population          55 non-null     int64
 2   Total               55 non-null     int64
 3   Violent             55 non-null     int64
 4   Property            55 non-null     int64
 5   Murder              55 non-null     int64
 6   Forcible_Rape       55 non-null     int64
 7   Robbery             55 non-null     int64
 8   Aggravated_assault  55 non-null     int64
 9   Burglary            55 non-null     int64
 10  Larceny_Theft       55 non-null     int64
 11  Vehicle_Theft       55 non-null     int64
dtypes: int64(12)
memory usage: 5.3 KB


In [153]:
us_data.dtypes

Year                  int64
Population            int64
Total                 int64
Violent               int64
Property              int64
Murder                int64
Forcible_Rape         int64
Robbery               int64
Aggravated_assault    int64
Burglary              int64
Larceny_Theft         int64
Vehicle_Theft         int64
dtype: object

In [154]:
us_data.describe()

Year    Population         Total       Violent      Property  \
count    55.00000  5.500000e+01  5.500000e+01  5.500000e+01  5.500000e+01   
mean   1987.00000  2.461556e+08  1.060315e+07  1.188806e+06  9.412499e+06   
std      16.02082  4.166216e+07  3.175931e+06  4.597107e+05  2.738512e+06   
min    1960.00000  1.793232e+08  3.384200e+06  2.884600e+05  3.095700e+06   
25%    1973.50000  2.106215e+08  9.096958e+06  9.253150e+05  8.060014e+06   
50%    1987.00000  2.422829e+08  1.140151e+07  1.322390e+06  1.018259e+07   
75%    2000.50000  2.833697e+08  1.308449e+07  1.432762e+06  1.160509e+07   
max    2014.00000  3.188571e+08  1.487290e+07  1.932270e+06  1.296110e+07   

             Murder  Forcible_Rape        Robbery  Aggravated_assault  \
count     55.000000      55.000000      55.000000        5.500000e+01   
mean   17317.236364   72714.800000  416449.890909        6.811818e+05   
std     4267.442664   28278.006893  150506.403782        2.960255e+05   
min     8530.000000   17190.000000  106670.000000        1.543200e+05   
25%    14819.000000   53400.000000  354911.500000        4.384300e+05   
50%    17030.000000   84230.000000  423557.000000        7.412910e+05   
75%    20561.500000   92930.000000  512137.000000        9.002150e+05   
max    24700.000000  109060.000000  687730.000000        1.135610e+06   

           Burglary  Larceny_Theft  Vehicle_Theft  
count  5.500000e+01   5.500000e+01   5.500000e+01  
mean   2.424017e+06   5.959947e+06   1.028614e+06  
std    7.259004e+05   1.846401e+06   3.455693e+05  
min    9.121000e+05   1.855400e+06   3.282000e+05  
25%    2.105336e+06   4.843350e+06   7.896260e+05  
50%    2.329950e+06   6.591900e+06   1.032200e+06  
75%    3.073600e+06   7.168450e+06   1.239302e+06  
max    3.795200e+06   8.142200e+06   1.661700e+06

In [155]:
#checking null values 
us_data.isnull().sum()

Year                  0
Population            0
Total                 0
Violent               0
Property              0
Murder                0
Forcible_Rape         0
Robbery               0
Aggravated_assault    0
Burglary              0
Larceny_Theft         0
Vehicle_Theft         0
dtype: int64

#### no null values present

In [156]:
us_data.head()

Year  Population    Total  Violent  Property  Murder  Forcible_Rape  \
0  1960   179323175  3384200   288460   3095700    9110          17190   
1  1961   182992000  3488000   289390   3198600    8740          17220   
2  1962   185771000  3752200   301510   3450700    8530          17550   
3  1963   188483000  4109500   316970   3792500    8640          17650   
4  1964   191141000  4564600   364220   4200400    9360          21420   

   Robbery  Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
0   107840              154320    912100        1855400         328200  
1   106670              156760    949600        1913000         336000  
2   110860              164570    994300        2089600         366800  
3   116470              174210   1086400        2297800         408300  
4   130390              203050   1213200        2514400         472800

In [158]:
#distinct number of years involved in data
pd.Categorical(us_data["Year"])

[1960, 1961, 1962, 1963, 1964, ..., 2010, 2011, 2012, 2013, 2014]
Length: 55
Categories (55, int64): [1960, 1961, 1962, 1963, ..., 2011, 2012, 2013, 2014]

### minimum Year: 1960 to maximum Year: 2014

### Which year has maximum Burglary 

In [171]:
gr = us_data.groupby("Year")
gr = gr.sum()
gr = gr.sort_values(["Burglary"], ascending = False)
gr.head()

Population     Total  Violent  Property  Murder  Forcible_Rape  Robbery  \
Year                                                                            
1980   225349264  13408300  1344520  12063700   23040          82990   565840   
1981   229146000  13423800  1361820  12061900   22520          82500   592910   
1982   231534000  12974400  1322390  11652000   21010          78770   553130   
1979   220099000  12249500  1208030  11041500   21460          76390   480700   
1975   213124000  11292400  1039710  10252700   20510          56090   470500   

      Aggravated_assault  Burglary  Larceny_Theft  Vehicle_Theft  
Year                                                              
1980              672650   3795200        7136900        1131700  
1981              663900   3779700        7194400        1087800  
1982              669480   3447100        7142500        1062400  
1979              629480   3327700        6601000        1112800  
1975              492620   3265300        5977700        1009600

# Third Data Set: master_data

#### https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/datasets.csv

In [173]:
master_data = pd.read_csv("https://raw.githubusercontent.com/vincentarelbundock/Rdatasets/master/datasets.csv")

In [174]:
master_data.head()

Package           Item                                              Title  \
0     AER        Affairs                   Fair's Extramarital Affairs Data   
1     AER   ArgentinaCPI                  Consumer Price Index in Argentina   
2     AER      BankWages                                         Bank Wages   
3     AER  BenderlyZwick  Benderly and Zwick Data: Inflation, Growth and...   
4     AER      BondYield                                    Bond Yield Data   

   Rows  Cols  n_binary  n_character  n_factor  n_logical  n_numeric  \
0   601     9         2            0         2          0          7   
1    80     2         0            0         0          0          2   
2   474     4         2            0         3          0          1   
3    31     5         0            0         0          0          5   
4    60     2         0            0         0          0          2   

                                                 CSV  \
0  https://vincentarelbundock.github.io/Rdatasets...   
1  https://vincentarelbundock.github.io/Rdatasets...   
2  https://vincentarelbundock.github.io/Rdatasets...   
3  https://vincentarelbundock.github.io/Rdatasets...   
4  https://vincentarelbundock.github.io/Rdatasets...   

                                                 Doc  
0  https://vincentarelbundock.github.io/Rdatasets...  
1  https://vincentarelbundock.github.io/Rdatasets...  
2  https://vincentarelbundock.github.io/Rdatasets...  
3  https://vincentarelbundock.github.io/Rdatasets...  
4  https://vincentarelbundock.github.io/Rdatasets...

In [176]:
#shape
master_data.shape

(1714, 12)

In [178]:
master_data.columns

Index(['Package', 'Item', 'Title', 'Rows', 'Cols', 'n_binary', 'n_character',
       'n_factor', 'n_logical', 'n_numeric', 'CSV', 'Doc'],
      dtype='object')

In [179]:
master_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1714 entries, 0 to 1713
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Package      1714 non-null   object
 1   Item         1714 non-null   object
 2   Title        1714 non-null   object
 3   Rows         1714 non-null   int64 
 4   Cols         1714 non-null   int64 
 5   n_binary     1714 non-null   int64 
 6   n_character  1714 non-null   int64 
 7   n_factor     1714 non-null   int64 
 8   n_logical    1714 non-null   int64 
 9   n_numeric    1714 non-null   int64 
 10  CSV          1714 non-null   object
 11  Doc          1714 non-null   object
dtypes: int64(7), object(5)
memory usage: 160.8+ KB


In [180]:
master_data.describe()

Rows         Cols     n_binary  n_character     n_factor  \
count  1.714000e+03  1714.000000  1714.000000  1714.000000  1714.000000   
mean   3.900833e+03    13.086348     1.931155     0.302800     1.270128   
std    3.997025e+04   171.966384    17.416499     1.266075     3.194779   
min    2.000000e+00     1.000000     0.000000     0.000000     0.000000   
25%    3.400000e+01     3.000000     0.000000     0.000000     0.000000   
50%    1.055000e+02     5.000000     0.000000     0.000000     0.000000   
75%    5.847500e+02     9.000000     2.000000     0.000000     2.000000   
max    1.414593e+06  6831.000000   624.000000    17.000000    64.000000   

         n_logical    n_numeric  
count  1714.000000  1714.000000  
mean      0.030922    11.429988  
std       0.405278   171.905339  
min       0.000000     0.000000  
25%       0.000000     2.000000  
50%       0.000000     3.000000  
75%       0.000000     7.000000  
max      11.000000  6830.000000

In [181]:
master_data.describe(include = "all")

Package     Item                       Title          Rows  \
count        1714     1714                        1714  1.714000e+03   
unique         53     1658                        1580           NaN   
top     Stat2Data  housing  Internal functions for gap           NaN   
freq          211        3                          18           NaN   
mean          NaN      NaN                         NaN  3.900833e+03   
std           NaN      NaN                         NaN  3.997025e+04   
min           NaN      NaN                         NaN  2.000000e+00   
25%           NaN      NaN                         NaN  3.400000e+01   
50%           NaN      NaN                         NaN  1.055000e+02   
75%           NaN      NaN                         NaN  5.847500e+02   
max           NaN      NaN                         NaN  1.414593e+06   

               Cols     n_binary  n_character     n_factor    n_logical  \
count   1714.000000  1714.000000  1714.000000  1714.000000  1714.000000   
unique          NaN          NaN          NaN          NaN          NaN   
top             NaN          NaN          NaN          NaN          NaN   
freq            NaN          NaN          NaN          NaN          NaN   
mean      13.086348     1.931155     0.302800     1.270128     0.030922   
std      171.966384    17.416499     1.266075     3.194779     0.405278   
min        1.000000     0.000000     0.000000     0.000000     0.000000   
25%        3.000000     0.000000     0.000000     0.000000     0.000000   
50%        5.000000     0.000000     0.000000     0.000000     0.000000   
75%        9.000000     2.000000     0.000000     2.000000     0.000000   
max     6831.000000   624.000000    17.000000    64.000000    11.000000   

          n_numeric                                                CSV  \
count   1714.000000                                               1714   
unique          NaN                                               1714   
top             NaN  https://vincentarelbundock.github.io/Rdatasets...   
freq            NaN                                                  1   
mean      11.429988                                                NaN   
std      171.905339                                                NaN   
min        0.000000                                                NaN   
25%        2.000000                                                NaN   
50%        3.000000                                                NaN   
75%        7.000000                                                NaN   
max     6830.000000                                                NaN   

                                                      Doc  
count                                                1714  
unique                                               1714  
top     https://vincentarelbundock.github.io/Rdatasets...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

In [182]:
#checking null values

master_data.isnull().sum()

Package        0
Item           0
Title          0
Rows           0
Cols           0
n_binary       0
n_character    0
n_factor       0
n_logical      0
n_numeric      0
CSV            0
Doc            0
dtype: int64

##### no null values present

In [184]:
master_data.dtypes

Package        object
Item           object
Title          object
Rows            int64
Cols            int64
n_binary        int64
n_character     int64
n_factor        int64
n_logical       int64
n_numeric       int64
CSV            object
Doc            object
dtype: object

In [190]:
#feature having object data type

master_data.dtypes[master_data.dtypes=="object"].index

Index(['Package', 'Item', 'Title', 'CSV', 'Doc'], dtype='object')

In [192]:
feat_obj = master_data[['Package', 'Item', 'Title', 'CSV', 'Doc']]
feat_obj.head()

Package           Item                                              Title  \
0     AER        Affairs                   Fair's Extramarital Affairs Data   
1     AER   ArgentinaCPI                  Consumer Price Index in Argentina   
2     AER      BankWages                                         Bank Wages   
3     AER  BenderlyZwick  Benderly and Zwick Data: Inflation, Growth and...   
4     AER      BondYield                                    Bond Yield Data   

                                                 CSV  \
0  https://vincentarelbundock.github.io/Rdatasets...   
1  https://vincentarelbundock.github.io/Rdatasets...   
2  https://vincentarelbundock.github.io/Rdatasets...   
3  https://vincentarelbundock.github.io/Rdatasets...   
4  https://vincentarelbundock.github.io/Rdatasets...   

                                                 Doc  
0  https://vincentarelbundock.github.io/Rdatasets...  
1  https://vincentarelbundock.github.io/Rdatasets...  
2  https://vincentarelbundock.github.io/Rdatasets...  
3  https://vincentarelbundock.github.io/Rdatasets...  
4  https://vincentarelbundock.github.io/Rdatasets...

In [196]:
#checking number of times different categories present in object data types

col = ['Package', 'Item', 'Title', 'CSV', 'Doc']

for i in col:
    print(f"\nColumn name: {i}, {feat_obj[i].value_counts()}")


Column name: Package, Stat2Data         211
openintro         198
Ecdat             134
DAAG              121
AER               107
MASS               87
datasets           84
stevedata          69
carData            63
boot               49
HistData           46
HSAUR              41
robustbase         40
fpp2               38
drc                35
causaldata         33
vcd                33
mosaicData         31
KMsurv             27
psych              23
COUNT              21
gap                18
pscl               17
tidyr              15
plm                14
survival           14
ISLR               13
ggplot2            11
lme4               11
geepack             9
cluster             9
evir                7
lattice             7
texmex              7
quantreg            7
mediation           6
gt                  6
rpart               6
sem                 6
nycflights13        5
dplyr               5
forecast            5
sandwich            4
HLMdiag             4
reshape2 

## Replacing Title with first word present in Title

In [206]:
title_len = len(feat_obj["Title"])

for i in range(0,title_len):
    feat_obj["Title"][i] = feat_obj["Title"][i].split(" ")[0]

C:\Users\USER\Anaconda3\envs\DeepLearning\lib\site-packages\IPython\core\interactiveshell.py:3418: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


In [207]:
feat_obj.head()

Package           Item     Title  \
0     AER        Affairs    Fair's   
1     AER   ArgentinaCPI  Consumer   
2     AER      BankWages      Bank   
3     AER  BenderlyZwick  Benderly   
4     AER      BondYield      Bond   

                                                 CSV  \
0  https://vincentarelbundock.github.io/Rdatasets...   
1  https://vincentarelbundock.github.io/Rdatasets...   
2  https://vincentarelbundock.github.io/Rdatasets...   
3  https://vincentarelbundock.github.io/Rdatasets...   
4  https://vincentarelbundock.github.io/Rdatasets...   

                                                 Doc  
0  https://vincentarelbundock.github.io/Rdatasets...  
1  https://vincentarelbundock.github.io/Rdatasets...  
2  https://vincentarelbundock.github.io/Rdatasets...  
3  https://vincentarelbundock.github.io/Rdatasets...  
4  https://vincentarelbundock.github.io/Rdatasets...

# Fourth Data Set: beer.txt

#### https://raw.githubusercontent.com/justmarkham/DAT8/master/data/beer.txt

In [208]:
beer_data = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/beer.txt", sep = "\t")

In [209]:
beer_data.head()

name calories sodium alcohol cost
0         Budweiser 144 15 4.7 0.43
1           Schlitz 151 19 4.9 0.43
2         Lowenbrau 157 15 0.9 0.48
3        Kronenbourg 170 7 5.2 0.73
4          Heineken 152 11 5.0 0.77

In [210]:
beer_data.columns

Index(['name calories sodium alcohol cost'], dtype='object')